src/main/java/org/nuxeo/binary/metadata/internals/BinaryMetadataServiceImpl.java

```diff
@@ -197,16 +198,25 @@ public void writeMetadata(DocumentModel doc, String mappingDescriptorId) {
             // Write doc properties from outputs.
             for (String metadata : blobMetadataOutput.keySet()) {
                 Object metadataValue = blobMetadataOutput.get(metadata);
+                boolean metadataIsArray = metadataValue.getClass().isArray() || metadataValue instanceof ArrayList;
                 String property = metadataMapping.get(metadata);
-                if (!(metadataValue instanceof Date) && !(metadataValue instanceof Collection)
-                        && !(metadataValue.getClass().isArray())) {
+                if (!(metadataValue instanceof Date) && !(metadataValue instanceof Collection) && !(metadataIsArray)) {
                     metadataValue = metadataValue.toString();
                 }
                 if (metadataValue instanceof String) {
                     // sanitize string for PostgreSQL textual storage
                     metadataValue = ((String) metadataValue).replace("\u0000", "");
                 }
                 try {
+                    if (doc.getProperty(property).isList()) {
+                        if (!metadataIsArray) {
+                            metadataValue = Arrays.asList(metadataValue);
+                        }
+                    } else {
+                        if (metadataIsArray) {
+                            metadataValue = metadataValue.toString();
```

Comment:
If `metadataValue` is an `Object[]` then `toString()` won't do what you want, you'll want to use `Arrays.toString((Object[]) metadataValue)` in that case. You should also add a unit test for that case.